### PR TITLE
Dev branch

### PR DESCRIPTION
**Pull request by szippy**

### CODE REVIEW

**Comment by mikegrima** [comment]
Hello!

The changes requested should be a separate plugin for Slack and not part of the `custom_alerter.py` file.

File: security_monkey/alerters/custom_alerter.py [hunk 1]
```diff
@@ -37,9 +40,40 @@ def report_auditor_changes(auditor):
...
+            alert_slack(auditor)
 
 
 def report_watcher_changes(watcher):
     for alerter_class in alerter_registry:
         alerter = alerter_class()
         alerter.report_watcher_changes(watcher)
+
+
+def alert_slack(auditor):
```
Comment (mikegrima) [inline security_monkey/alerters/custom_alerter.py]: Please move to a custom alerter Slack plugin.

File: security_monkey/alerters/custom_alerter.py [hunk 2]
```diff
@@ -37,9 +40,40 @@ def report_auditor_changes(auditor):
...
+        for issue in item.confirmed_new_issues:
+            x = {
+                'text': "New Issue: Index: {!s}\n Account: {!s}\n Region: {!s}\n Name: {!s}".format(item.index, item.account, item.region, item.name)
+            }
+
+            slack_config['attachments'].append(x)
+            slack_post(slack_config, slack_hook)
+
+
+def slack_post(slack_config, slack_hook):
```
Comment (mikegrima) [inline security_monkey/alerters/custom_alerter.py]: Same

File: security_monkey/alerters/custom_alerter.py [hunk 3]
```diff
@@ -37,9 +40,40 @@ def report_auditor_changes(auditor):
     for alerter_class in alerter_registry:
         alerter = alerter_class()
         alerter.report_auditor_changes(auditor)
+        if app.config.get('SLACK_HOOK'):
```
Comment (mikegrima) [inline security_monkey/alerters/custom_alerter.py]: Please move to a separate slack plugin.

**Review by mikegrima — CHANGES_REQUESTED** [review]
Please move the code out of the standard alerter file and into a dedicated slack plugin.

**Comment by szippy** [comment]
Moved all slack alert logic out to slack_alerter.py

File: security_monkey/alerters/custom_alerter.py [hunk 4]
```diff
@@ -21,6 +21,7 @@
 
 """
 from security_monkey import app
+from slack_alerter import alert_slack
```
Comment (mikegrima) [inline security_monkey/alerters/custom_alerter.py]: Remove

File: security_monkey/alerters/custom_alerter.py [hunk 5]
```diff
@@ -37,6 +38,7 @@ def report_auditor_changes(auditor):
     for alerter_class in alerter_registry:
         alerter = alerter_class()
         alerter.report_auditor_changes(auditor)
+        alert_slack(auditor)
```
Comment (mikegrima) [inline security_monkey/alerters/custom_alerter.py]: Remove

File: security_monkey/alerters/slack_alerter.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+"""alert slack with any auditor"""
+
+import json
+
+import requests
+from security_monkey import app
+
+
+def alert_slack(auditor):
```
Comment (mikegrima) [inline security_monkey/alerters/slack_alerter.py]: This should be a subclass of the `AlerterType` https://github.com/Netflix/security_monkey/blob/develop/security_monkey/alerters/custom_alerter.py

This code should be implemented within the constructor for the class.

File: security_monkey/alerters/slack_alerter.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            for issue in item.confirmed_new_issues:
+                x = {
+                    'text': "New Issue: Index: {!s}\n Account: {!s}\n Region: {!s}\n Name: {!s}".format(item.index, item.account, item.region, item.name)
+                }
+
+                slack_config['attachments'].append(x)
+                slack_post(slack_config, slack_hook)
+
+
+def slack_post(slack_config, slack_hook):
```
Comment (mikegrima) [inline security_monkey/alerters/slack_alerter.py]: This should be within the `report_auditor_changes` method.

**Review by mikegrima — CHANGES_REQUESTED** [review]

**Comment by szippy** [comment]
Made requested changes according to example in security_monkey/misc.md

File: security_monkey/alerters/slack_alerter.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+"""alert slack with any auditor"""
+
+import json
+
+import requests
+from security_monkey import app
+from security_monkey.alerters import custom_alerter
+
+alerter_registry = []
```
Comment (mikegrima) [inline security_monkey/alerters/slack_alerter.py]: Move these to the `SlackAlerter`'s constructor `__init__`

File: security_monkey/alerters/slack_alerter.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+slack_config = {}
+slack_hook = ""
+
+
+class SlackAlerter(object):
+    __metaclass__ = custom_alerter.AlerterType
+
+    if app.config.get('SLACK_HOOK'):
```
Comment (mikegrima) [inline security_monkey/alerters/slack_alerter.py]: Make a constructor, and move this code to the `__init__`.

File: security_monkey/alerters/slack_alerter.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+                    'text': "New Issue: Index: {!s}\n Account: {!s}\n Region: {!s}\n Name: {!s}".format(item.index, item.account, item.region, item.name)
+                }
+
+                slack_config['attachments'].append(x)
+                try:
+                    requests.post(slack_hook, data=json.dumps(slack_config))
+                    slack_config['attachments'] = []
+
+                except Exception as e:
+                    print("something has gone wrong with the slack post. Please check your configuration. " + e)
```
Comment (mikegrima) [inline security_monkey/alerters/slack_alerter.py]: Please make use of the built in Logger.

```
from security_monkey import app
...
except Exception as e:
    app.logger.exception(e)
    app.logger.error("something has gone wrong with the slack post. Please check your configuration. " + e)
```

**Review by mikegrima — CHANGES_REQUESTED** [review]

**Comment by szippy** [comment]
Created __init__, moved variable declarations into __init__ method, and swapped print statement out for logger.

File: security_monkey/alerters/slack_alerter.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+            self.slack_config = {
+                'channel': app.config.get('SLACK_CHANNEL'),
+                'username': app.config.get('SLACK_USERNAME'),
+                'icon_emoji': app.config.get('SLACK_ICON'),
+            }
+
+            self.slack_hook = app.config.get('SLACK_HOOK')
+
+            self.slack_config['attachments'] = []
```
Comment (mikegrima) [inline security_monkey/alerters/slack_alerter.py]: Remove this.

File: security_monkey/alerters/slack_alerter.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+    def report_auditor_changes(self, auditor):
+        for item in auditor.items:
+            for issue in item.confirmed_new_issues:
+                x = {
+                    'text': "New Issue: Index: {!s}\n Account: {!s}\n Region: {!s}\n Name: {!s}".format(item.index, item.account, item.region, item.name)
+                }
+
+                self.slack_config['attachments'].append(x)
```
Comment (mikegrima) [inline security_monkey/alerters/slack_alerter.py]: Make `attachments` a local variable, and then this should be good to go 👍

**Review by mikegrima — CHANGES_REQUESTED** [review]
Almost done :)

File: security_monkey/alerters/slack_alerter.py (hunk 12 above)
Comment (szippy) [reply security_monkey/alerters/slack_alerter.py]: attachments needs to be part of slack_config to be posted correctly into the slack hook. I set it equal to the message, and removed it from the init method.

Comment (mikegrima) [reply security_monkey/alerters/slack_alerter.py]: Makes sense.

**Review by mikegrima — APPROVED** [review]

**Pull request merged**